pkg/cri/resource-manager/policy/builtin/balloons/balloons-policy.go

```diff
@@ -0,0 +1,941 @@
+func (p *balloons) ExportResourceData(c cache.Container) map[string]string {
+	return nil
+}
+
+// Introspect provides data for external introspection.
+func (p *balloons) Introspect(*introspect.State) {
+	return
+}
+
+// balloonByContainer returns a balloon that contains a container.
+func (p *balloons) balloonByContainer(c cache.Container) *Balloon {
+	podID := c.GetPodID()
+	cID := c.GetCacheID()
+	for _, bln := range p.balloons {
+		for _, ctrID := range bln.PodIDs[podID] {
+			if ctrID == cID {
+				return bln
+			}
+		}
+	}
+	return nil
+}
+
+// balloonsByPod returns balloons that contain any container of a pod.
+func (p *balloons) balloonsByPod(pod cache.Pod) []*Balloon {
+	podID := pod.GetID()
+	blns := []*Balloon{}
+	for _, bln := range p.balloons {
+		if _, ok := bln.PodIDs[podID]; ok {
+			blns = append(blns, bln)
+		}
+	}
+	return blns
+}
+
+// balloonsByDef returns list of balloons instantiated from a balloon
+// definition.
+func (p *balloons) balloonsByDef(blnDef *BalloonDef) []*Balloon {
+	balloons := []*Balloon{}
+	for _, balloon := range p.balloons {
+		if balloon.Def == blnDef {
+			balloons = append(balloons, balloon)
+		}
+	}
+	return balloons
+}
+
+// balloonDefByName returns a balloon definition with a name.
+func (p *balloons) balloonDefByName(defName string) *BalloonDef {
+	for _, blnDef := range p.bpoptions.BalloonDefs {
+		if blnDef.Name == defName {
+			return blnDef
+		}
+	}
+	return nil
+}
+
+// balloonDefByNamespace returns a balloon definition with a namespace
+// in its Namespaces list.
+func (p *balloons) balloonDefByNamespace(namespace string) *BalloonDef {
+	for _, blnDef := range p.bpoptions.BalloonDefs {
+		for _, blnNamespace := range blnDef.Namespaces {
+			if blnNamespace == namespace {
+				return blnDef
+			}
+		}
+	}
+	return nil
+}
+
+func (p *balloons) chooseBalloonDef(am AssignMethod, c cache.Container) (*BalloonDef, error) {
+	switch am {
+	case AssignNamespace:
+		namespace := c.GetNamespace()
+		if blnDef := p.balloonDefByNamespace(namespace); blnDef != nil {
+			return blnDef, nil
+		}
+	case AssignNamespaceMust:
+		namespace := c.GetNamespace()
+		if blnDef := p.balloonDefByNamespace(namespace); blnDef != nil {
+			return blnDef, nil
+		}
+		return nil, balloonsError("no balloon for namespace %q required by assign method %s", namespace, am)
+	case AssignAnnotation:
+		if blnDefName, ok := c.GetEffectiveAnnotation(balloonKey); ok {
+			if blnDef := p.balloonDefByName(blnDefName); blnDef != nil {
+				return blnDef, nil
+			}
+		}
+	case AssignAnnotationMust:
+		if blnDefName, ok := c.GetEffectiveAnnotation(balloonKey); ok {
+			if blnDef := p.balloonDefByName(blnDefName); blnDef != nil {
+				return blnDef, nil
+			}
+			return nil, balloonsError("no balloon for annotation %q required by assign method %s", blnDefName, am)
+		}
+	case AssignSystemToReserved:
+		if c.GetNamespace() == metav1.NamespaceSystem {
+			return p.balloons[0].Def, nil
+		}
```

Comment:
Cannot we change things so that
 - we always create a balloon from the reserved CPUs on startup and associate it with the `kube-system` namespace
 - let this case be handled by an `AssignNamespace[Must]`
For that to work, we might need a bit of extra post-config logic, to make sure that `AssignNamespace` is there in the configuration, etc.

It would feel so much less error-prone and easier to understand/configure to me to somehow get rid of `AssignSystemToReserved` altogether...